File: wrapper/android/src/main/java/com/zeticai/mlange/pipeline/ZeticMLangePipeline.kt
package com.zeticai.mlange.pipeline

import com.zeticai.mlange.feature.PipelineFeature
import com.zeticai.mlange.inputsource.InputSource

class ZeticMLangePipeline<T>(
    private val feature: PipelineFeature<T>,
    private val inputSource: InputSource? = null,
) {

    fun run(input: Array<ByteArray>): T {
        return feature.run(input)
    }

    fun loop(frame: (T) -> Unit) {
        inputSource?.acquire {
            val output = run(arrayOf(it))
            frame(output)
        }
    }
}

File: wrapper/ios/Pipeline/ZeticMLangePipeline.swift
import Foundation
import Combine

public class ZeticMLangePipeline<Input, Output>: ObservableObject {
    private let feature: any PipelineFeature<Input, Output>
    private let inputSource: (any InputSource<Input>)?
    
    @Published public var latestResult: Output?
    
    public init(
        feature: any PipelineFeature<Input, Output>,
        inputSource: (any InputSource<Input>)? = nil
    ) {
        self.feature = feature
        self.inputSource = inputSource
    }
    
    public func run(input: Input) -> Output? {
        return feature.run(input)
    }
    
    public func startLoop() {
        inputSource?.acquire { [weak self] input in
            DispatchQueue.main.async {
                self?.latestResult = self?.run(input: input)
            }
        }
    }
    
    public func stopLoop() {
        inputSource?.stop()
    }
}
